src/BufferManager.cpp

```diff
@@ -0,0 +1,104 @@
+/**
+ * @file BufferManager.cpp
+ *
+ * This is part of the DUNE DAQ Application Framework, copyright 2021.
+ * Licensing/copyright details are in the COPYING file that you should have
+ * received with this code.
+ */
+#include "trigger/BufferManager.hpp"
+
+#include <iostream>
+
+namespace dunedaq::trigger {
+
+BufferManager::BufferManager(long unsigned int buffer_size)
+  : m_buffer_max_size(buffer_size)
+  , m_buffer_earliest_start_time(0)
+  , m_buffer_latest_end_time(0)
+{
+
+}
+
+BufferManager::~BufferManager()
+{
+
+}
+
+bool
+BufferManager::add(trigger::TPSet& tps)
+{
+  if(m_tpset_buffer.size() >= m_buffer_max_size) //delete oldest TPSet if buffer full
+  {
+    auto firstIt = m_tpset_buffer.begin();
+    m_tpset_buffer.erase(firstIt);
+    // add some warning message here, saying that buffer max capacity was reached?
+  }
+  if( (m_buffer_earliest_start_time == 0) || (tps.start_time < m_buffer_earliest_start_time) )
+    m_buffer_earliest_start_time = tps.start_time;
+  
+  if( (m_buffer_latest_end_time == 0) || (tps.end_time > m_buffer_latest_end_time) )
+    m_buffer_latest_end_time = tps.end_time;
+
+  return m_tpset_buffer.insert(tps).second; //false if tps with same start_time already exists
+}
+
+std::vector<trigger::TPSet>
+BufferManager::get_tpsets_in_window(dataformats::timestamp_t start_time, dataformats::timestamp_t end_time)
+{
+  std::vector<trigger::TPSet> tpsets_output;
+
+  if(end_time < m_buffer_earliest_start_time)
+  {
+    // add warning here saying data requested doesn't exist in the buffer anymore
+    return tpsets_output;
+  }
+
+  if(start_time > m_buffer_latest_end_time) //condition (4)
+  {
+    // add warning here saying data requested hasn't arrived in the buffer yet
+    // need to creat a queue of "pending" data request. How?
+    return tpsets_output;
+  }
+
+  for(auto& tps: m_tpset_buffer)
```

Comment:
For now it's fine to loop over all of the TPSets to find the ones we want. We may in future want to try to speed this up by noting that `std::set` is stored sorted, so we can only loop over the range of TPSets in the window by using `lower_bound` and `upper_bound` to find the relevant bounds. For large buffers, this is likely to make a performance difference

Reply:
Thanks! I've update it accordingly.